proposals/multiple_user_script_worlds.md

```diff
@@ -0,0 +1,285 @@
+    /**
+     * Specifies which pages this user script will be injected into. See
+     * Match Patterns for more details on the syntax of these strings. This
+     * property must be specified for ${ref:register}.
+     */
+    matches?: string[];
+    /**
+     * Specifies when JavaScript files are injected into the web page. The
+     * preferred and default value is document_idle
+     */
+    runAt?: RunAt;
+    /**
+     * The JavaScript execution environment to run the script in. The default
+     * is `USER_SCRIPT`
+     */
+    world?: ExecutionWorld;
+
++   /**
++    * If specified, specifies a specific user script world ID to execute in.
++    * Only valid if `world` is omitted or is `USER_SCRIPT`. If omitted, the
++    * script will execute in the default user script world.
++    */
++   worldId?: string;
+  }
+}
+```
+
+Note that the signatures of the related functions, including `configureWorld()`,
+`register()`, and others are left unchanged.
+
+When the developer specifies a `worldId` in either the `WorldProperties` or
+a `RegisteredUserScript`, the browser will create a separate user script world
+for those cases. If `worldId` is omitted, a "default" user script world will
+be used.
+
+### New Permissions
+
+No new permissions are necessary. This is inline with the `userScripts` API's
+current functionality and purpose.
+
+### Manifest File Changes
+
+No manifest file changes are necessary.
+
+## Security and Privacy
+
+### Exposed Sensitive Data
+
+This does not expose any additional sensitive data.
+
+### Abuse Mitigations
+
+This API does not provide any additional avenue for abuse beyond what the
+`userScripts` API is already capable of.
+
+Extension developers may abuse this API by requiring too many user script
+worlds, which would have a heavy performance cost. However, this is not a
+security consideration. Additionally, browsers can mitigate this by enforcing
+a limit of the maximum number of user script worlds an extension may register
+or have active at a given time.
+
+These enforcements will be left up to the browser to implement, if they feel
+necessary.
+
+### Additional Security Considerations
+
+No additional security considerations. This API may (mildly) increase security
+by (somewhat) isolating individual user scripts from interacting with one
+another.
+
+## Alternatives
+
+### Existing Workarounds
+
+Today, user script managers go to various lengths to try to isolate behavior
+between user scripts. This includes having various frozen types, making heavy
+use of function closures, ensuring injection before untrusted code, and more.
+
+These workarounds are fragile and require significantly more complex code in
+user script managers, while providing a lesser guarantee of isolation than a
+separate user script world would provide.
+
+### Open Web API
+
+The concept of user scripts should not be exposed to the web; this should not
+be an open web API.
+
+## Implementation Notes
+
+N/A
+
+## Future Work
```

Comment:
Would we ever expose `worldId` to the content side? A use case that I can see is enabling (synchronous) cross-world communication or execution. I've wanted to create such a proposal for a while.

Reply:
By "content" I meant in the renderer. Because doing so opens two things: firstly the ability to do interworld communication and two, enabling extensions to use it as a hack to encode a world-specific configuration in the string.

Since you included the mention of the first, the answer appears to be yes and I'll resolve this thread.